src/tcache.c

```diff
@@ -795,6 +794,8 @@ tcaches_create(tsd_t *tsd, base_t *base, unsigned *r_ind) {
 
 	bool err;
 
+	malloc_mutex_lock(tsd_tsdn(tsd), &tcaches_mtx);
+
 	if (tcaches_create_prep(tsd, base)) {
```

Comment:
It looks there's no simple way to 1) not holding the lock while allocating, and 2) reserving the tcache id and handling OOM properly at the same time?

Reply:
Yeah; the other thing we could do is check and the recheck once we reacquire the lock. But that's actually reasonably complex, because it then means that the tcache allocation becomes speculative and we might need to do cleanup for it as well. (Also, tcache-sized object allocation is fast enough that it's not clear that doubling the number of acquisitions/releases is worth it to reduce the hold time).